src/pool/srv_pool.c

```diff
@@ -2320,11 +2320,25 @@ bulk_cb(const struct crt_bulk_cb_info *cb_info)
 	return 0;
 }
 
-void
-ds_pool_connect_handler(crt_rpc_t *rpc)
+/* Currently we only maintain compatibility between 2 versions */
+#define NUM_POOL_VERSIONS	2
+
+/*
+ * Max supported pool version that server could handle from client pool
+ * connection.
+ */
+static unsigned int max_supported_pool_version[NUM_POOL_VERSIONS] = {
+	/* for client pool layout version DAOS_PROP_PO_GLOBAL_VERSION */
+	DAOS_PROP_PO_GLOBAL_VERSION,
+	/* for client pool layout version DAOS_PROP_PO_GLOBAL_VERSION - 1 */
+	1,
```

Comment:
Instead of a numeric literal, shouldn't this be `DAOS_PROP_PO_GLOBAL_VERSION - 1`?

Makes me wonder if the array is the right way to represent this. Essentially we want to make calculations based on the global version to determine supported versions, assuming they are sequential as shown here.

Reply:
No, it could be sometimes DAOS_PROP_PO_GLOBAL_VERSION or DAOS_PROP_PO_GLOBAL_VERSION - 1, depends on if there are some in-compatible change which could make client unable to read data from newer pool.

For example, in 2.0/2.2, it is compatible, so they are both 1. but for 2.2/2.4, it will be [2, 1], and maybe in the 2.4/2.6 it might be [3, 2]

Reply:
Ah I didn't notice this issue again in this most recent patch. Shouldn't the code be using DAOS_POOL_GLOBAL_VERSION  and DAOS_POOL_GLOBAL_VERSION  -1  (2 and 1)

(and not DAOS_PROP_PO_GLOBAL_VERSION that is a very different value: 14)
?

Reply:
Good catch on Ken's part. I am still trying to understand this though. Do we want to be chained to updating this array for every release, depending on whether the pool version has been bumped between the latest and previous? It just seems highly likely to be missed or forgotten.

To me it seems practical to do something a bit more generic that isn't coupled so tightly to release versions, e.g. support the current and previous pool version regardless of what the release version says. If we truly want to couple to the release version we should probably be more explicit about it.

Reply:
I agree this is likely to be missed, i've added comments in the DAOS_POOL_GLOBAL_VERSION to update supported array when you bumped it.

And unfortunately, i did not get way to fix this easily, for example when upgrading DAOS pool from 2.2 to 2.4 with EC rotation enabled, we need upgrade client version as well to access latest pool format, in this case, latest pool version won't work for old clients, then old client should be upgraded to latest version to access new format.

We could not easily say we always support current and previous pool version regardless of what the release version says.

But I guess a safe way is to always reject old clients connection to new pools, that means 2.2 client connect to 2.4 pool won't work, 2.4 client connects to 2.6 pool version won't work.(unless you upgrade client to latest as well). This is reliable but not that friendly, eg 2.0 client actually access 2.2 pool without problem.

And i guess in the future, most of time,  older client could connect new pool without problem, upgrade client as well means we change some fundamental client prototype, i guess this won't happen frequently, so i'd prefer to keep the array now.

What do you think?
Thanks!

Reply:
A rethink on this, i think maybe it is fine to reject old clients connection with newer pool format is fine.
This is safe and simple for now.

For allowed old client connect to old pool only.

Reply:
I think that is fine too. Assuming we are meaning it is determined by the pool version not the release version.

Reply:
My understanding from a discussion with @johannlombardi was that old client software should be (generally) able to connect to newer server/engine software. The 2.2 client connecting to 2.4 (with pool format updated for EC rotation) I guess is an exception to the general goal. 

So is the plan for now (with this patch) to simply reject older client software attempting to connect to newer pools? And decide to address the general goal later in some future patch?

Reply:
I think after applying PR, old clients still could connect newer servers if that is old pool, except is if pool is upgraded or newly created, old clients won't connect, @johannlombardi  are you ok with this? or do you think we'd better maintain supported matric which will try best to support old clients connecting new pool if it is possible.